Added CDK/TypeScript folder which contains CDKv2 code.

Anirudh1612 opened the pull request:
Added CDK/TypeScript folder which contains CDKv2 code to provision lambda power tuner.

alexcasalboni commented on cdk/typescript/template.yaml:
```diff
@@ -0,0 +1,123 @@
+Transform: AWS::Serverless-2016-10-31
```
Is this meant to be included in the repo? Isn't it a generated template?

alexcasalboni replied:
This should have disappeared, right? 😄

alexcasalboni commented on cdk/typescript/test/the-lambda-power-tuner.test.ts:
```diff
@@ -0,0 +1,16 @@
+  expectCDK(stack).to(haveResourceLike("AWS::Serverless::Application", {
+    "Location": {
+      "ApplicationId": "arn:aws:serverlessrepo:us-east-1:451282441545:applications/aws-lambda-power-tuning",
+      "SemanticVersion": "4.2.0"
```
This looks like a test that will fail pretty often, as we increase the version number :) I'd suggest we make it version independent, otherwise we'll need to update this file every time a new version is published

alexcasalboni commented on cdk/typescript/lib/the-lambda-power-tuner-stack.ts:
```diff
@@ -0,0 +1,41 @@
+    let lambdaResource = "*";
+
+    // A lambda function to use to test the powertuner
+    let exampleLambda = new lambda.Function(this, 'lambdaHandler', {
```
Personally, I think this is also optional :) you don't necessarily have to define the Lambda function you're power-tuning in the same stack. Actually, most people just deploy Lambda Power Tuning once and then use it with as many functions as needed. 

I understand this is a sample pattern and customers will be able to customize it as needed, but I'd try to make it very clear that they don't have to deploy their functions in the same stack.

alexcasalboni reviewed (changes requested):
Thanks for the PR! 🙏 

I've left a few comments to address :)

Anirudh1612 commented:
@alexcasalboni I have updated the the-lambda-power-tuner.test.ts file. It was not reflecting CDKv2, updated and tested the new code and it works fine now. Also I have removed README.md file. Will review the other changes suggested and push a new commit later.

alexcasalboni commented on .gitignore:
```diff
@@ -22,3 +22,6 @@ layer-sdk/src
 samconfig.toml
+
+#template file 
+template.yml
```
Why this? We do need to track the main template file :)

alexcasalboni replied:
Did you mean to include this in the `cdk/typescript/.gitignore` file?

Anirudh1612 replied:
> Did you mean to include this in the `cdk/typescript/.gitignore` file?

Yes, I will add this file there.

alexcasalboni commented on cdk/typescript/cdk.json:
```diff
@@ -0,0 +1,64 @@
+      "test"
+    ]
+  },
+  "context": {
```
What is this for? Do we need all this EC2/ECS/RDS stuff?

Anirudh1612 replied:
> What is this for? Do we need all this EC2/ECS/RDS stuff?

This file points NPM to execute the "bin" file whenever a CDK command is executed.

alexcasalboni replied:
> > What is this for? Do we need all this EC2/ECS/RDS stuff?
> 
> This file points NPM to execute the "bin" file whenever a CDK command is executed.

I understand, but why do we need all the EC2/ECS/RDS stuff? Is it required for this CDK project?

alexcasalboni commented on cdk/typescript/img/state-machine-success.png:
This screenshot is obsolete now. Can you please take a new one of the latest UI?

Anirudh1612 replied:
Sure

alexcasalboni commented on cdk/typescript/img/well_architected.png:
Do we need this Well-Architected image at all?

Anirudh1612 replied:
No, will remove this image.

alexcasalboni commented on cdk/typescript/lib/the-lambda-power-tuner-stack.ts:
```diff
@@ -0,0 +1,26 @@
+    super(scope, id, props);
+
+    let powerValues = '128,256,512,1024,1536,3008';
+    let lambdaResource = "*";
```
I would suggest that we include (commented out) all the possible parameters. Similar to what we do here for the CloudFormation template: https://github.com/alexcasalboni/aws-lambda-power-tuning/blob/master/scripts/deploy-sar-app.yml#L13

alexcasalboni replied:
Can you please make the indentation consistent here?

alexcasalboni commented on cdk/typescript/test/the-lambda-power-tuner.test.ts:
```diff
@@ -0,0 +1,17 @@
+import * as cdk from 'aws-cdk-lib';
+import * as TheLambdaPowerTuner from '../lib/the-lambda-power-tuner-stack';
+
+test('SAR Application Created', () => {
```
I assume this runs smoothly when you execute `npm run test`, right?

I cannot think of additional tests to run here for now, but things will definitely change if/once we drop SAR support :)

Anirudh1612 replied:
Yes, I have tested this out. It works smoothly.

alexcasalboni reviewed (changes requested):
A few more comments :)

alexcasalboni commented on cdk/typescript/README.md:
```diff
@@ -0,0 +1,83 @@
+# Deploy The Lambda Power Tuner with CDK
```
I'm wondering if this file should be moved up one folder? (for all the CDK projects across different languages)

See the C# PR here too: https://github.com/alexcasalboni/aws-lambda-power-tuning/pull/234

alexcasalboni commented on cdk/typescript/lib/the-lambda-power-tuner-stack.ts:
```diff
@@ -0,0 +1,39 @@
+    new sam.CfnApplication(this, 'powerTuner', {
+      location: {
+        applicationId: 'arn:aws:serverlessrepo:us-east-1:451282441545:applications/aws-lambda-power-tuning',
+        semanticVersion: '4.2.0'
```
Please use the lates version here -> `4.3.4`

alexcasalboni reviewed (changes requested):
Still a few comments to address :)

alexcasalboni reviewed (approved):
Ok, we're ready to merge this 🚀
